src/cls/rgw/cls_rgw.cc

```diff
@@ -640,19 +631,41 @@ int rgw_bucket_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
 		__func__, escape_str(kiter->first).c_str());
         continue;
       }
+      CLS_LOG(20, "%s: working on key=%s len=%zu",
+	      __func__, kiter->first.c_str(), kiter->first.size());
+      start_after_omap_key = kiter->first;
 
-      if (!entry.is_valid()) {
-        CLS_LOG(20, "%s: entry %s[%s] is not valid",
-		__func__, key.name.c_str(), key.instance.c_str());
-        continue;
-      }
+      // these checks are performed also by RGW
+      if (rgw_do_osd_side_checks) {
+        rgw_bucket_dir_entry entry;
+        try {
+          const bufferlist& entrybl = kiter->second;
+          auto eiter = entrybl.cbegin();
+          decode(entry, eiter);
+        } catch (ceph::buffer::error& err) {
+          CLS_LOG(1, "ERROR: %s: failed to decode entry, key=%s",
+                  __func__, kiter->first.c_str());
+          return -EINVAL;
+        }
 
-      // filter out noncurrent versions, delete markers, and initial marker
-      if (!op.list_versions &&
-	  (!entry.is_visible() || op.start_obj.name == key.name)) {
-        CLS_LOG(20, "%s: entry %s[%s] is not visible",
-		__func__, key.name.c_str(), key.instance.c_str());
-        continue;
+        start_after_entry_key = entry.key;
+
+        if (!entry.is_valid()) {
+          CLS_LOG(20, "%s: entry %s[%s] is not valid",
+                  __func__, key.name.c_str(), key.instance.c_str());
+          continue;
+        }
+
+        // filter out noncurrent versions, delete markers, and initial marker
+        if (!op.list_versions &&
+            (!entry.is_visible() || op.start_obj.name == key.name)) {
+          CLS_LOG(20, "%s: entry %s[%s] is not visible",
+                  __func__, key.name.c_str(), key.instance.c_str());
+          continue;
+        }
+      } else if (!start_after_entry_key.empty()) {
+	// this will be filled from decoding the very last entry
+        start_after_entry_key = cls_rgw_obj_key{};
       }
 
       if (has_delimiter) {
```

Comment:
Hmm, perhaps the straight case (no delimiter, no extensive versioning) of bucket listing could bypass the `cls_rgw` entirely – RGW (or maybe even `cls_rgw_client`) could emit `CEPH_OSD_OP_OMAPGETVALS`  (instead of `CEPH_OSD_OP_CALL`) relying the `is_valid()` and `is_visible()` checks are performed client-side.

The question is how large the _ugly namespace_ of keys can be.

Dynamic switching between the direct and cls-assisted paths is not unthinkable.